### PR TITLE
Improve is_port_free

### PR DESCRIPTION
**Pull request by ldoktor**
Hello guys, it took me quite long to find out our is_port_free function is seriously insufficient (first I blamed qemu, but it's us). The first commit fixes the issue where ipv4 address is already freed, but ipv6 address is still occupied, the second commit makes sure we don't allow using port that is already used by UDP (not sure why but at least on my system it allows that without complaint). Third replaces "localhost" with "", because on "localhost" python binds only to 127.0.0.1/::1 but on "" it binds to all localhost interfaces (like 192.168.122.1, ...), which is commonly used by programs and we were not detectin these. After that follows fixes and optimization.

PS: This (hopefully) fixes (or at least decreases probability) very rare Avocado-vt migration issue where ports are not free even though we claim they are.

### CODE REVIEW

**Comment by ldoktor** [comment]
@clebergnu would you please take a look at the approach and let me know what parts are acceptable so I can fix only the parts that are considerable for merging?

**Comment by ldoktor** [comment]
OK I played with it a bit more and the last 2 patches can be relaced with bind to "" or "0.0.0.0", which does the same job (binds to all interfaces). Anyway take a look at the first 2 patches but they are really required as following passes:

```python
>>> import socket
>>> a=socket.socket(socket.AF_INET, socket.SOCK_STREAM)
>>> b=socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
>>> a.bind(("", 5200))
>>> b.bind(("", 5200))
```

**Comment by ldoktor** [comment]
@clebergnu there was no review yet so I force-pushed a new, simplified, version. It does not require extra dependency on Avocado side (only in requirements-selftest) and adds some cleanups. All of those fixes are required to avoid reporting free ports that might be used by ordinary programs (still it won't check for custom protocols, but this should cover the majority).

File: selftests/unit/test_utils_network.py [hunk 1]
```diff
@@ -33,5 +34,39 @@ def test_release_port(self):
         self.assertNotIn(22, tracker.retained_ports)
 
 
+class FreePort(unittest.TestCase):
+
+    def test_is_port_free(self):
+        port = network.find_free_port()
+        self.assertTrue(network.is_port_free(port, "localhost"))
```
Comment (clebergnu) [inline selftests/unit/test_utils_network.py]: I don't claim to have a perfect solution for this, but this (as is `network.find_free_port()`) is susceptible to race conditions if other software is running on the same system.

One possibility, though, that just crossed my mind, is the fact that we can use any address on 127.0.0.1/8 range, and something like that can be also true for IPv6.  In theory, if we pick an address other than 127.0.0.1, we at least, minimize a lot the possibility of clashes and thus race conditions.

File: avocado/utils/network.py [hunk 2]
```diff
@@ -25,31 +25,36 @@
 
 #: Families taken into account in this class
 FAMILIES = (socket.AF_INET, socket.AF_INET6)
+#: Protocols taken into account in this class
+PROTOCOLS = (socket.SOCK_STREAM, socket.SOCK_DGRAM)
```
Comment (clebergnu) [inline avocado/utils/network.py]: Same comment here about adding an optional parameter for protocols (as for families)

File: requirements-selftests.txt [hunk 3]
```diff
@@ -34,3 +34,6 @@ lxml==4.2.4
 # in make develop. The proper solution would be migrate
 # our make develop process to use pip
 libvirt-python==4.6.0
+
+# For avocado.utils.network selftests
+netifaces
```
Comment (clebergnu) [inline requirements-selftests.txt]: We need to either make the test optional on the presence or not of the `netifaces` module, or add it as a hard build requirement also on the RPM spec file.

File: avocado/utils/network.py [hunk 4]
```diff
@@ -23,29 +23,37 @@
...
-            s.connect((address, port))
-            free = False
-    except socket.error:
-        if address == "localhost":
-            free = False
-        else:
-            free = True
-    s.close()
-    return free
+        for family in FAMILIES:
```
[the remark this answers is not in the document]
Comment (ldoktor) [reply avocado/utils/network.py]: I don't think it's that necessary. Until now nobody really cared and complained so let's wait until some user requires it...

**Pull request closed without merging**